File: py/packages/genkit/src/genkit/_ai/_tools.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tool-specific types and utilities for the Genkit framework."""

import inspect
import json
from collections.abc import Callable
from contextvars import ContextVar
from typing import Any, cast, get_type_hints

from opentelemetry import trace as trace_api
from pydantic import BaseModel

from genkit._core._action import Action, ActionKind, ActionRunContext, _is_context_annotation
from genkit._core._error import GenkitError, GenkitInterrupt
from genkit._core._logger import get_logger
from genkit._core._middleware import GenerateMiddlewareContext
from genkit._core._registry import Registry
from genkit._core._typing import ToolDefinition, ToolRequest, ToolRequestPart, ToolResponse, ToolResponsePart

logger = get_logger(__name__)


class Tool:
    """A registered tool: a callable handle backed by an :class:`~genkit._core._action.Action`.

    Obtain instances via :func:`define_tool`, :func:`define_interrupt`, :func:`tool`, or the
    ``@ai.tool`` decorator rather than constructing directly.
    """

    def __init__(self, action: Action) -> None:
        self._action = action

    @property
    def name(self) -> str:
        """Tool name (registry key)."""
        return self._action.name

    @property
    def description(self) -> str:
        """Human-readable description sent to the model."""
        return self._action.description or ''

    @property
    def input_schema(self) -> dict[str, object] | None:
        """JSON Schema for the tool's input, as sent on the wire."""
        return self._action.input_schema

    @property
    def output_schema(self) -> dict[str, object] | None:
        """JSON Schema for the tool's output."""
        return self._action.output_schema

    def definition(self) -> ToolDefinition:
        """Return the wire-format ToolDefinition for this tool."""
        return ToolDefinition(
            name=self.name,
            description=self.description,
            input_schema=self.input_schema,
            output_schema=self.output_schema,
        )

    def action(self) -> Action:
        """Return the underlying :class:`~genkit._core._action.Action` registered for this tool."""
        return self._action

    async def __call__(self, *args: Any, **kwargs: Any) -> Any:  # noqa: ANN401
        """Run the tool and return the unwrapped response value."""
        return (await self._action.run(*args, **kwargs)).response


# Context variables for propagating resumed metadata to tools
_tool_resumed_metadata: ContextVar[dict[str, Any] | None] = ContextVar('tool_resumed_metadata', default=None)
# Stashed copy of tool_request.input when restart replaces input (JSON; shape is per tool).
_tool_original_input: ContextVar[Any | None] = ContextVar('tool_original_input', default=None)  # noqa: ANN401


class ToolRunContext(ActionRunContext):
    """Tool execution context with interrupt support."""

    def __init__(
        self,
        ctx: ActionRunContext,
        resumed_metadata: dict[str, Any] | None = None,
        original_input: Any = None,  # noqa: ANN401 - prior tool_request.input when replacing on restart
    ) -> None:
        """Initialize from parent ActionRunContext.

        Args:
            ctx: Parent action context
            resumed_metadata: Metadata from previous interrupt (if resumed)
            original_input: Original tool input before replacement (if resumed)
        """
        super().__init__(
            context=ctx.context,
            streaming_callback=ctx.streaming_callback,
            abort_signal=ctx.abort_signal,
        )
        self.resumed_metadata = resumed_metadata
        self.original_input = original_input

    def is_resumed(self) -> bool:
        """Return True if this execution is resuming after an interrupt."""
        return self.resumed_metadata is not None


class Interrupt(GenkitInterrupt):  # noqa: N818 - public Genkit name; not renamed *Error for style
    """Exception for interrupting tool execution with user-facing API.

    Raise ``Interrupt(metadata)`` from a tool or from tool middleware (e.g. ``wrap_tool``).
    Exceptions from ``tool.run`` are wrapped in GenkitError
    with ``cause=Interrupt``; generation attaches interrupt metadata to the pending tool
    request.

    To resume, use ``respond_to_interrupt`` or ``restart_tool``.
    """

    def __init__(self, metadata: dict[str, Any] | None = None) -> None:
        """Initialize an Interrupt exception.

        Args:
            metadata: Attached to the tool request on the wire. Use a plain dict; for a
                Pydantic model, pass ``m.model_dump(mode="json")``.
        """
        super().__init__()
        self.metadata: dict[str, Any] = {} if metadata is None else metadata
        if self.metadata:
            span = trace_api.get_current_span()
            if span.is_recording():
                try:
                    span.set_attribute('genkit:metadata:interrupt', json.dumps(self.metadata))
                except Exception:
                    span.set_attribute('genkit:metadata:interrupt', str(self.metadata))


def _tool_response_part(
    interrupt: ToolRequestPart,
    output: Any,  # noqa: ANN401 - arbitrary tool/interrupt reply payload (JSON)
    metadata: dict[str, Any] | None = None,
) -> ToolResponsePart:
    """Build a ``ToolResponsePart`` for an interrupted tool request (interrupt reply channel)."""
    interrupt_metadata = metadata if metadata is not None else True
    tool_req = interrupt.tool_request
    return ToolResponsePart(
        tool_response=ToolResponse(
            ref=tool_req.ref,
            name=tool_req.name,
            output=output,
        ),
        metadata={'interruptResponse': interrupt_metadata},
    )


def respond_to_interrupt(
    response: Any,  # noqa: ANN401 - user reply or tool output for resume_respond
    *,
    interrupt: ToolRequestPart,
    metadata: dict[str, Any] | None = None,
) -> ToolResponsePart:
    """Build a ``ToolResponsePart`` for a pending tool interrupt.

    Pass the return value to ``generate(..., resume_respond=interrupt_response)``.

    Args:
        response: Tool output / user reply for this interrupt.
        interrupt: The interrupted ``ToolRequestPart`` (e.g. from ``response.interrupts``).
        metadata: Optional metadata for the interrupt response channel.
    """
    return _tool_response_part(interrupt, response, metadata)


def restart_tool(
    *,
    interrupt: ToolRequestPart,
    replace_input: Any | None = None,  # noqa: ANN401 - new tool input; shape is per tool
    resumed_metadata: dict[str, Any] | None = None,
) -> ToolRequestPart:
    """Build a restart ``ToolRequestPart`` for a pending tool interrupt.

    Pass the return value to ``generate(..., resume_restart=...)``.

    Args:
        interrupt: The interrupted ``ToolRequestPart`` (e.g. from ``response.interrupts``).
        replace_input: Optional new ``tool_request.input`` for this run (previous input is
            stored in ``metadata.replacedInput`` when this is set).
        resumed_metadata: Passed to the tool as ``ToolRunContext.resumed_metadata``.

    Returns:
        A ``ToolRequestPart`` for ``resume_restart`` / message history.

    Example:
        ``restart_tool(interrupt=trp, resumed_metadata={"tool_approved": True})``
    """
    tool_req = interrupt.tool_request
    new_meta: dict[str, Any] = dict(interrupt.metadata or {})

    new_meta['resumed'] = resumed_metadata if resumed_metadata is not None else True

    new_input = tool_req.input
    if replace_input is not None:
        new_meta['replacedInput'] = tool_req.input
        new_input = replace_input

    return ToolRequestPart(
        tool_request=ToolRequest(
            name=tool_req.name,
            ref=tool_req.ref,
            input=new_input,
        ),
        metadata=new_meta,
    )


def _resume_context_from_tool_request_part(
    tool_request_part: ToolRequestPart,
) -> tuple[dict[str, Any] | None, Any | None]:
    """Read resume/restart fields from a tool request part's metadata."""
    meta = tool_request_part.metadata or {}
    raw_resumed = meta.get('resumed')
    if raw_resumed is True:
        resumed_meta: dict[str, Any] | None = {}
    elif isinstance(raw_resumed, dict):
        resumed_meta = raw_resumed
    else:
        resumed_meta = None

    original_input = meta.get('replacedInput')
    return resumed_meta, original_input


async def run_tool_request(
    *,
    tool: Action,
    tool_request_part: ToolRequestPart,
    ctx: GenerateMiddlewareContext | None = None,
) -> Any:  # noqa: ANN401 - tool output follows registered handler
    """Execute a tool request with generate-scoped context and resume metadata.

    Pipes ``GenerateMiddlewareContext.custom_context`` and ``telemetry_labels``
    into ``tool.run``, and sets resume ContextVars from ``tool_request_part``
    metadata so ``ToolRunContext`` reflects ``resumed`` / ``replacedInput``.
    """
    resumed_meta, original_input = _resume_context_from_tool_request_part(tool_request_part)
    token_meta = _tool_resumed_metadata.set(resumed_meta)
    token_input = _tool_original_input.set(original_input)
    run_context = dict(ctx.custom_context) if ctx and ctx.custom_context else None
    telemetry_labels = cast(dict[str, object], dict(ctx.telemetry_labels)) if ctx and ctx.telemetry_labels else None
    try:
        return (
            await tool.run(
                tool_request_part.tool_request.input,
                context=run_context,
                telemetry_labels=telemetry_labels,
                abort_signal=ctx.abort_signal if ctx else None,
            )
        ).response
    finally:
        _tool_resumed_metadata.reset(token_meta)
        _tool_original_input.reset(token_input)


def restart_interrupt_error(interrupt: Interrupt) -> GenkitError:
    """Build the FAILED_PRECONDITION error for an Interrupt raised during tool restart.

    A tool cannot raise another interrupt while it is being restarted. Include the underlying
    interrupt reason (e.g. ToolApproval's ``Tool not in approved list: ...``) so the
    error points at missing approval metadata instead of sounding like a missing SDK feature.
    """
    metadata = interrupt.metadata
    if isinstance(metadata, dict):
        reason = metadata.get('message')
    elif isinstance(metadata, str):
        # Defensive: Interrupt is typed as dict metadata, but a plain string
        # argument would land here and must not AttributeError on .get().
        reason = metadata
    else:
        reason = None
    if isinstance(reason, str) and reason.strip():
        message = f'Tool interrupted again during restart: {reason}'
    else:
        message = 'Tool interrupted again during restart.'
    return GenkitError(status='FAILED_PRECONDITION', message=message, cause=interrupt)


async def run_tool_after_restart(
    *,
    tool: Action,
    restart_trp: ToolRequestPart,
    ctx: GenerateMiddlewareContext | None = None,
) -> ToolResponsePart:
    """Run a tool for ``resume_restart``: applies ``resumed`` / ``replacedInput`` from metadata.

    Sets the same context variables as the tool wrapper so ToolRunContext reflects
    a resumed run. A tool cannot raise another interrupt while it is being restarted.
    """
    try:
        tool_response = await run_tool_request(tool=tool, tool_request_part=restart_trp, ctx=ctx)
    except (GenkitError, Interrupt) as e:
        intr = (
            e.cause
            if isinstance(e, GenkitError) and isinstance(e.cause, Interrupt)
            else (e if isinstance(e, Interrupt) else None)
        )
        if intr is not None:
            logger.debug(
                'restarted tool triggered an interrupt',
                tool=restart_trp.tool_request.name,
            )
            raise restart_interrupt_error(intr) from e
        raise

    return ToolResponsePart(
        tool_response=ToolResponse(
            name=restart_trp.tool_request.name,
            ref=restart_trp.tool_request.ref,
            output=tool_response.model_dump() if isinstance(tool_response, BaseModel) else tool_response,
        )
    )


def _get_func_description(func: Callable[..., Any], description: str | None = None) -> str:
    """Return description if provided, otherwise use the function's docstring."""
    if description is not None:
        return description
    if func.__doc__ is not None:
        return func.__doc__
    return ''


def _define_tool(
    registry: Registry,
    func: Callable[..., Any],
    name: str | None = None,
    description: str | None = None,
    *,
    input_schema: type[BaseModel] | dict[str, object] | None = None,
) -> Tool:
    """Register a function as a tool.

    Normally, the input_schema and output_schema are inferred from func. However,
    in some cases, like define_interrupt, the app developer doesn't have a way to
    express the input schema in the func signature.

    In that case, the app developer can pass in an input_schema to override the inferred schema.
    This will ensure that the model requesting the tool will see the correct input shape.
    """
    if not inspect.iscoroutinefunction(func):
        raise TypeError(f'Tool function must be async. Got sync function: {getattr(func, "__name__", repr(func))}')

    tool_name = name if name is not None else getattr(func, '__name__', None)
    if tool_name is None:
        raise ValueError(f'Cannot infer a tool name from {func!r}; pass name= explicitly.')
    tool_description = _get_func_description(func, description)

    input_spec = inspect.getfullargspec(func)

    try:
        resolved_annotations = get_type_hints(func)
    except (NameError, TypeError, AttributeError):
        resolved_annotations = input_spec.annotations
    arg_types = [resolved_annotations.get(arg, Any) for arg in input_spec.args]

    async def tool_fn_wrapper(*args: Any) -> Any:  # noqa: ANN401 - arity dispatch; args/return follow registered tool
        # Record resumed metadata on the current span for observability.
        resumed_meta = _tool_resumed_metadata.get()
        if resumed_meta:
            span = trace_api.get_current_span()
            if span.is_recording():
                try:
                    span.set_attribute('genkit:metadata:resumed', json.dumps(resumed_meta))
                except Exception:
                    span.set_attribute('genkit:metadata:resumed', str(resumed_meta))
        original_input = _tool_original_input.get()

        def tool_run_context(action_ctx: ActionRunContext) -> ToolRunContext:
            return ToolRunContext(
                action_ctx,
                resumed_metadata=resumed_meta,
                original_input=original_input,
            )

        # Dynamic dispatch by arity; payload types follow the registered tool (not expressible here).
        match len(input_spec.args):
            case 0:
                return await func()
            case 1:
                if _is_context_annotation(arg_types[0]):
                    return await func(tool_run_context(cast(ActionRunContext, args[0])))
                return await func(args[0])
            case 2:
                return await func(
                    args[0],
                    tool_run_context(cast(ActionRunContext, args[1])),
                )
            case _:
                raise ValueError('tool must have 0-2 args...')

    action = registry.register_action(
        name=tool_name,
        kind=ActionKind.TOOL,
        description=tool_description,
        fn=tool_fn_wrapper,
        metadata_fn=func,
    )
    if input_schema is not None:
        action._override_input_schema(input_schema)

    return Tool(action)


def define_tool(
    registry: Registry,
    func: Callable[..., Any],
    name: str | None = None,
    description: str | None = None,
    *,
    input_schema: type[BaseModel] | dict[str, object] | None = None,
) -> Tool:
    """Register a function as a tool.

    Tool input/output JSON Schemas are inferred from ``func`` (first parameter and return type).

    Args:
        registry: The registry to register the tool in.
        func: The async function to register as a tool. Must be a coroutine function.
        name: Optional name for the tool. Defaults to the function name.
        description: Optional description. Defaults to the function's docstring.
        input_schema: Optional input schema override (Pydantic model or JSON-schema dict).

    Raises:
        TypeError: If func is not an async function.
    """
    return _define_tool(registry, func, name, description, input_schema=input_schema)


def tool(
    func: Callable[..., Any],
    *,
    name: str | None = None,
    description: str | None = None,
    input_schema: type[BaseModel] | dict[str, object] | None = None,
) -> Tool:
    """Define an ephemeral tool for a single ``generate`` call.

    Unlike ``@ai.tool()``, this does not register the tool on the app, so it
    cannot be referenced by name later. Use for dynamic tools that exist only
    for one call.

    Args:
        func: Async tool implementation (same 0–2 argument rules as :func:`define_tool`).
        name: Tool name for the model. Defaults to ``func.__name__``.
        description: Sent to the model. Defaults to the function docstring.
        input_schema: Optional input schema override (Pydantic model or JSON-schema dict).

    Raises:
        TypeError: If ``func`` is not a coroutine function.
        ValueError: If no ``name`` is given and ``func`` has no ``__name__``.

    Example:
        from genkit import tool

        async def lookup_price(sku: str) -> str:
            return '12.00'

        ephemeral = tool(lookup_price, name='lookup_price')
        res = await ai.generate(prompt='Price of SKU-1?', tools=[ephemeral])
    """
    return _define_tool(Registry(), func, name, description, input_schema=input_schema)


def define_interrupt(
    registry: Registry,
    name: str,
    *,
    description: str | None = None,
    request_metadata: dict[str, Any] | Callable[[Any], dict[str, Any]] | None = None,  # noqa: ANN401
    input_schema: type[BaseModel] | dict[str, object] | None = None,
) -> Tool:
    """Register a tool that always interrupts execution.

    An interrupt tool is a special tool that always raises ``Interrupt`` with
    optional metadata. This is useful for explicit human-in-the-loop checkpoints.
    For tools that sometimes run logic and sometimes interrupt, use ``define_tool``
    and raise ``Interrupt`` from the handler (or use ``ToolRunContext``).

    Args:
        registry: The registry to register the interrupt tool in
        name: Tool name (registry key)
        description: Tool description shown to the model
        request_metadata: Static metadata dict or ``(input) -> dict`` for the interrupt
        input_schema: Optional wire input schema (Pydantic model or JSON schema dict). The
            interrupt handler is typed as ``Any``; pass this so the model sees a concrete shape.

    Returns:
        The registered tool callable (same shape as ``define_tool``).

    Example:
        def get_meta(input: dict) -> dict:
            return {"action": input.get("action"), "requires_approval": True}

        confirm = define_interrupt(
            registry,
            "confirm",
            description="Requires user approval",
            request_metadata=get_meta,
        )
    """

    async def interrupt_wrapper(input: Any) -> Any:  # noqa: ANN401 - wire JSON args; never returns (raises Interrupt)
        # Interrupt tools accept arbitrary JSON args like any tool.
        meta = None
        if callable(request_metadata):
            meta = request_metadata(input)
        elif request_metadata is not None:
            meta = request_metadata
        raise Interrupt(meta)

    return _define_tool(
        registry,
        interrupt_wrapper,
        name=name,
        description=description,
        input_schema=input_schema,
    )

File: go/ai/exp/tools.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package exp

import (
	"context"
	"errors"
	"fmt"

	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/ai/exp/tool"
	"github.com/firebase/genkit/go/core/api"
	"github.com/firebase/genkit/go/internal/base"
)

// ToolFunc is the function signature for tools created with [DefineTool] and [NewTool].
type ToolFunc[In, Out any] = func(ctx context.Context, input In) (Out, error)

// InterruptibleToolFunc is the function signature for tools created with
// [DefineInterruptibleTool] and [NewInterruptibleTool]. The resumed parameter
// is non-nil when the tool is being re-executed after an interrupt.
type InterruptibleToolFunc[In, Out, Resume any] = func(ctx context.Context, input In, res *Resume) (Out, error)

// Tool wraps an [ai.Tool] with experimental x package features
// such as a plain [context.Context] function signature and [tool.AttachParts].
//
// DEPRECATED(breaking): With breaking changes, Tool would not wrap ai.ToolDef.
// It would be the primary tool type, backed directly by core.DefineAction,
// eliminating the inner field and all delegation methods below.
type Tool[In, Out any] struct {
	inner *ai.ToolDef[In, *ai.MultipartToolResponse] // DEPRECATED(breaking): remove wrapper; Tool owns the action directly.
}

// DEPRECATED(breaking): All methods below exist only to implement ai.Tool by
// delegating to the wrapped ai.ToolDef. With breaking changes, Tool would own
// the action directly and implement these natively without delegation.

// Name returns the name of the tool.
func (t *Tool[In, Out]) Name() string { return t.inner.Name() }

// Definition returns the [ai.ToolDefinition] for this tool.
func (t *Tool[In, Out]) Definition() *ai.ToolDefinition { return t.inner.Definition() }

// RunRaw runs the tool with raw input.
func (t *Tool[In, Out]) RunRaw(ctx context.Context, input any) (any, error) {
	return t.inner.RunRaw(ctx, input)
}

// RunRawMultipart runs the tool with raw input and returns the full multipart response.
func (t *Tool[In, Out]) RunRawMultipart(ctx context.Context, input any) (*ai.MultipartToolResponse, error) {
	return t.inner.RunRawMultipart(ctx, input)
}

// Respond creates a tool response part for an interrupted tool request.
func (t *Tool[In, Out]) Respond(toolReq *ai.Part, outputData any, opts *ai.RespondOptions) *ai.Part {
	return t.inner.Respond(toolReq, outputData, opts)
}

// Restart creates a restart part using the legacy [ai.RestartOptions].
//
// DEPRECATED(breaking): Remove entirely. Superseded by [InterruptibleTool.Resume].
func (t *Tool[In, Out]) Restart(toolReq *ai.Part, opts *ai.RestartOptions) *ai.Part {
	return t.inner.Restart(toolReq, opts)
}

// Register registers the tool with the given registry.
func (t *Tool[In, Out]) Register(r api.Registry) { t.inner.Register(r) }

// InterruptibleTool is a [Tool] that supports typed interrupt/resume.
// The Res type parameter is the type of data the caller sends back when
// resuming the tool after an interrupt.
type InterruptibleTool[In, Out, Res any] struct {
	Tool[In, Out]
}

// Resume creates a restart part for resuming this interrupted tool with typed data.
// The data will be deserialized into the *Res parameter of the tool function
// when it is re-executed.
//
// Unlike [tool.Resume], this method also validates that the interrupted part
// belongs to this tool.
func (t *InterruptibleTool[In, Out, Resume]) Resume(part *ai.Part, res Resume) (*ai.Part, error) {
	if part == nil || !part.IsInterrupt() {
		return nil, fmt.Errorf("Resume: part is not an interrupted tool request")
	}
	if part.ToolRequest.Name != t.Name() {
		return nil, fmt.Errorf("Resume: tool request is for %q, not %q", part.ToolRequest.Name, t.Name())
	}
	return tool.Resume(part, res)
}

// Respond creates a tool response [ai.Part] for an interrupted tool request.
// Instead of re-executing the tool (as [Resume] does), this provides a
// pre-computed result directly.
//
// Unlike [tool.Respond], this method validates that the interrupted part
// belongs to this tool and accepts a strongly-typed output.
func (t *InterruptibleTool[In, Out, Resume]) Respond(part *ai.Part, output Out) (*ai.Part, error) {
	if part == nil || !part.IsInterrupt() {
		return nil, fmt.Errorf("Respond: part is not an interrupted tool request")
	}
	if part.ToolRequest.Name != t.Name() {
		return nil, fmt.Errorf("Respond: tool request is for %q, not %q", part.ToolRequest.Name, t.Name())
	}
	return tool.Respond(part, output)
}

// DefineTool creates a new tool with a simple function signature and registers it.
// The function receives a plain [context.Context] instead of [ai.ToolContext].
// Use [tool.AttachParts] inside the function to return additional content parts.
func DefineTool[In, Out any](
	r api.Registry,
	name, description string,
	fn ToolFunc[In, Out],
	opts ...ai.ToolOption,
) *Tool[In, Out] {
	t := NewTool(name, description, fn, opts...)
	t.Register(r)
	return t
}

// NewTool creates a new unregistered tool with a simple function signature.
// Use [tool.AttachParts] inside the function to return additional content parts.
func NewTool[In, Out any](
	name, description string,
	fn ToolFunc[In, Out],
	opts ...ai.ToolOption,
) *Tool[In, Out] {
	// DEPRECATED(breaking): Call core.NewAction directly instead of wrapping ai.NewMultipartTool.
	inner := ai.NewMultipartTool(name, description, wrapSimpleFunc(fn), opts...)
	return &Tool[In, Out]{inner: inner}
}

// DefineInterruptibleTool creates a new interruptible tool and registers it.
// The resumed parameter is non-nil when the tool is being resumed after an
// interrupt. Use [tool.Interrupt] inside the function to interrupt execution
// and send data to the caller.
func DefineInterruptibleTool[In, Out, Res any](
	r api.Registry,
	name, description string,
	fn InterruptibleToolFunc[In, Out, Res],
	opts ...ai.ToolOption,
) *InterruptibleTool[In, Out, Res] {
	t := NewInterruptibleTool(name, description, fn, opts...)
	t.Register(r)
	return t
}

// NewInterruptibleTool creates a new unregistered interruptible tool.
func NewInterruptibleTool[In, Out, Res any](
	name, description string,
	fn InterruptibleToolFunc[In, Out, Res],
	opts ...ai.ToolOption,
) *InterruptibleTool[In, Out, Res] {
	// DEPRECATED(breaking): Call core.NewAction directly instead of wrapping ai.NewMultipartTool.
	inner := ai.NewMultipartTool(name, description, wrapInterruptibleFunc(fn), opts...)
	return &InterruptibleTool[In, Out, Res]{Tool: Tool[In, Out]{inner: inner}}
}

// DEPRECATED(breaking): wrapSimpleFunc exists to adapt our func(context.Context, In) (Out, error)
// to ai.MultipartToolFunc[In] (which takes *ai.ToolContext). With breaking changes,
// core.DefineAction would accept our function signature directly, and the ToolContext
// adapter, resumed/originalInput extraction from ToolContext, and interrupt error
// conversion would all be unnecessary.
func wrapSimpleFunc[In, Out any](fn ToolFunc[In, Out]) ai.MultipartToolFunc[In] {
	return func(tc *ai.ToolContext, input In) (*ai.MultipartToolResponse, error) {
		ctx := tc.Context
		ctx, collector := tool.NewPartsContext(ctx)
		if tc.OriginalInput != nil {
			ctx = tool.SetOriginalInput(ctx, tc.OriginalInput)
		}

		output, err := fn(ctx, input)
		if err != nil {
			return nil, convertInterruptError(tc, err)
		}

		resp := &ai.MultipartToolResponse{Output: output}
		if parts := collector(); len(parts) > 0 {
			resp.Content = parts
		}
		return resp, nil
	}
}

// DEPRECATED(breaking): Same as wrapSimpleFunc — exists only to bridge between
// the new function signature and ai.MultipartToolFunc/ai.ToolContext.
func wrapInterruptibleFunc[In, Out, Resume any](fn InterruptibleToolFunc[In, Out, Resume]) ai.MultipartToolFunc[In] {
	return func(tc *ai.ToolContext, input In) (*ai.MultipartToolResponse, error) {
		ctx := tc.Context
		ctx, collector := tool.NewPartsContext(ctx)
		if tc.OriginalInput != nil {
			ctx = tool.SetOriginalInput(ctx, tc.OriginalInput)
		}

		// DEPRECATED(breaking): Resumed data would come from context keys set by
		// the generate loop directly, not from ai.ToolContext.Resumed.
		var res *Resume
		if tc.Resumed != nil {
			r, err := base.MapToStruct[Resume](tc.Resumed)
			if err != nil {
				return nil, fmt.Errorf("aix.wrapInterruptibleFunc: failed to convert resumed data: %w", err)
			}
			res = &r
		}

		output, err := fn(ctx, input, res)
		if err != nil {
			return nil, convertInterruptError(tc, err)
		}

		resp := &ai.MultipartToolResponse{Output: output}
		if parts := collector(); len(parts) > 0 {
			resp.Content = parts
		}
		return resp, nil
	}
}

// DEPRECATED(breaking): convertInterruptError exists because tool.InterruptError
// must be converted to ai's unexported toolInterruptError (via tc.Interrupt) for
// the generate loop to recognize it. With breaking changes, the generate loop
// would recognize tool.InterruptError directly.
func convertInterruptError(tc *ai.ToolContext, err error) error {
	var ie *tool.InterruptError
	if errors.As(err, &ie) {
		m, mapErr := toMap(ie.Data)
		if mapErr != nil {
			return fmt.Errorf("tool.Interrupt: failed to convert data: %w", mapErr)
		}
		return tc.Interrupt(&ai.InterruptOptions{Metadata: m})
	}
	return err
}

// DEPRECATED(breaking): toMap exists only for convertInterruptError above.
func toMap(v any) (map[string]any, error) {
	if v == nil {
		return nil, nil
	}
	if m, ok := v.(map[string]any); ok {
		return m, nil
	}
	return base.StructToMap(v)
}
